Implements support for ECDSA keys. Fixes #2163

atombrella opened the pull request:
I don't really know what stalled the other PR; I updated it a bit, moved `CHANGELOG` entries to the top. And then also tweaked a couple of tests. It's unclear to me if 256 and 384 should be the only two allowed key sizes for EC; I added 521, but can remove it again. Anyway, I took it up again, and really hope this time that the 3rd active PR to get support for EC keys will be merged. I'll do my best to address feedback promptly. I had a look at the design document that was referenced in #7247 and I believe everything was already addressed? Please correct me if I'm wrong. Since @pahrohfit did the work, and I only updated it a bit, he's added in `Co-Authored-By`. It could be reversed, if desired.

Some big websites have adopted elliptic curve cryptography, and I hope it'll become more widespread once `certbot` will support it. https://smallstep.com/certificate-manager/ supports it. It's similar to certbot in some ways, although it works also behind a firewall, in an internal network.

`echo | openssl s_client -connect yahoo.com:443 2>/dev/null | openssl x509 -text -noout | grep "Public-Key"`

## Pull Request Checklist

- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Include your name in `AUTHORS.md` if you like.

adferrand commented:
Thanks @atombrella!

Basically what stalled #7247 are the following things, which also applies to your PR:
* the PR is very close to the specification we agreed, the only thing to change essentially is to switch from `--ec-key-size` to `--elliptic-curve` and use SECG name of the curves, instead of bit size
* integration tests need to be added, as I said here: https://github.com/certbot/certbot/pull/7247#discussion_r321684775

@pahrohfit was not responding anymore, so if you are willing to continue the effort, that is great !

At this time sadly there is a blocker that will make the integration tests fail: some renewal parameters disappear with some association of CLI flags, like the key size, and so here also the key type and the curve for ECDSA. It is described here https://github.com/certbot/certbot/issues/7694, and a PR is under work here https://github.com/certbot/certbot/pull/7798.

atombrella commented:
> Thanks @atombrella!
> 
> Basically what stalled #7247 are the following things, which also applies to your PR:
> 
> * the PR is very close to the specification we agreed, the only thing to change essentially is to switch from `--ec-key-size` to `--elliptic-curve` and use SECG name of the curves, instead of bit size
> * integration tests need to be added, as I said here: [#7247 (comment)](https://github.com/certbot/certbot/pull/7247#discussion_r321684775)

OK. I have some unpushed changes to address the elliptic curve name feature. `openssl ec_params -list-curves` lists quite a few curves. Currently, I've limited it to names starting with `sec`, and bit sizes >= 256, by having it in those names in a list in . Any feedback here?

```python
    helpful.add(
        "security", "--elliptic-curve", type=str, choices=[
              'sect283k1',
              'sect283r1',
              'sect409k1',
              'sect409r1',
              'sect571k1',
              'sect571r1',
              'secp256k1',
              'secp384r1',
              'secp521r1',
        ], metavar="N",
        default=flag_default("elliptic_curve"), help=config_help("elliptic_curve"))
```

Integration tests are also a work-in-progress.

It's a good way also for me to brush up on this topic :)

> At this time sadly there is a blocker that will make the integration tests fail: some renewal parameters disappear with some association of CLI flags, like the key size, and so here also the key type and the curve for ECDSA. It is described here #7694, and a PR is under work here #7798.

OK. I'll have a closer look at those two issues/PRs.

atombrella commented on certbot/certbot/tests/testdata/README:
```diff
@@ -2,6 +2,13 @@ The following command has been used to generate test keys:
 
 	for x in 256 512 2048; do openssl genrsa -out rsa${k}_key.pem $k; done
 
+and for the elliptic curves, this command was used:
```
I'm not really sure if this is needed.

adferrand replied:
What do you mean here?

atombrella replied:
I'm not really doing anything with these keys, so I'm slightly confused about this part. 

I think it should `for x in "prime256v1" "secp384r1" "secp521r1"`, and then it's fine?

atombrella commented on certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:
```diff
@@ -421,19 +439,87 @@ def test_reuse_key(context):
+    context.certbot([
+        'certonly',
+        '-d', certname,
+        '--key-type ecdsa', '--elliptic-curve', 'secp384r1',
```
`--key-type` is not a valid argument, and I don't really understand why. A bit of help would be nice :)

adferrand replied:
It is because you pass `"--key-type ecdsa"` as a whole argument, literally an argument with a space character in it. What you want to pass is a flag and its value, which count as two arguments.

You can achieve that either by passing two arguments, `"--key-type"` and `"ecdsa"` in the list given to the exec function or use the form `"--key-type=ecdsa"`

atombrella commented on certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:
```diff
@@ -421,19 +439,87 @@ def test_reuse_key(context):
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    key2 = join(context.config_dir, 'archive', certname, 'privkey2.pem')
+    assert_elliptic_key(key2, SECP384R1)
```
I added a new `assert` for this operation, because I used it a few times. I think it's the only way to actually test the type of key used.

adferrand replied:
Makes sense. In this case, could you put the new assertion in the `certbot_integration_tests.certbot_tests.assertions` module? On top of the fact the module logically gather all custom assertions, they are also properly registered as assertions for pytest using `pytest.register_assert_rewrite` call done in `__init__`. It allows pytest to optimize output when these assertions functions raise an `AssertionError`.

atombrella commented on certbot/CHANGELOG.md:
```diff
@@ -24,6 +24,10 @@ More details about these changes can be found on our GitHub repo.
   using `update_account --register-unsafely-without-email`
+* CLI flag `--key-type` has been added to specify 'rsa' or 'ecdsa' (default 'rsa').
+  Only accepts a single value at this time.
+* CLI flag `--elliptic-curve` has been added which takes an NIST/SECG elliptic curve,
```
This needs to be made more precise. From what I read, only `secp256k1`, `secp384r1` and `secp521r1` are supported by the different browser vendors (probably defined in a standard).

osirisinferi replied:
> probably defined in a standard

The CA/Browser Forum Baseline Requirements to be exact. From "6.1.5 Key sizes":

> For ECDSA key pairs, the CA SHALL:
> •  Ensure that the key represents a valid point on the NIST P-256, NIST P-384 or NIST P-521 elliptic curve.

The naming of the curves is a little bit difficult though. For example, for every key size there are two curves: one with parameters associated with a Koblitz curve (denoted by the `k` in the curve name) and one with verifiably random parameters (denoted by the `r` in the curve name). As we can see from the [SEC 2: Recommended Elliptic Curve Domain Parameters (Version 2.0)](https://www.secg.org/sec2-v2.pdf) Table 2, the NIST recommended curves are the **random** curves. I.e., with a `r` in the `secp` cuve name. Unfortunately, you're not going to find the random parameter key size 256 by its "secp" name in OpenSSL. As we can read in [RFC 5480](https://tools.ietf.org/html/rfc5480#section-2.1.1.1), for some reason the `secp256r1` curve is referred to as `prime256v1` (also in OpenSSL and in the [Current Mozilla Security/Server Side TLS Wiki document](https://wiki.mozilla.org/index.php?title=Security/Server_Side_TLS&oldid=1229478#Intermediate_compatibility_.28recommended.29).

TL;DR: sed 's/secp256k1/secp256r1/' :wink:

atombrella replied:
Ah. Thanks :) I think I replaced it consistently, and also noticed I had `secp521k1` somewhere in my changes.

adferrand replied:
Nothing to add. Indeed these curves are the recommended ones, and the overall design we agreed for ECDSA keys in Certbot was to follow Mozilla recommandations here.

atombrella commented on certbot/README.rst:
```diff
@@ -67,7 +67,7 @@ Let's Encrypt Website: https://letsencrypt.org
 Community: https://community.letsencrypt.org
 
-ACME spec: `RFC 8555 <https://tools.ietf.org/html/rfc8555>`_
+ACME spec: https://ietf-wg-acme.github.io/acme/
```
Probably a result of a rebase. I'll revert it.

atombrella commented on certbot/certbot/_internal/cli/__init__.py:
```diff
@@ -309,6 +309,22 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):
         "security", "--rsa-key-size", type=int, metavar="N",
         default=flag_default("rsa_key_size"), help=config_help("rsa_key_size"))
+    helpful.add(
+        "security", "--elliptic-curve", type=str, choices=[
```
Note to self, this should only be a list of three curves.

adferrand replied:
I think it is the case now.

atombrella replied:
Yes. It wasn't when I wrote the comment. I'll mark it as resolved :)

atombrella commented on certbot/certbot/plugins/storage.py:
```diff
@@ -106,7 +106,7 @@ def put(self, key, value):
             self._initialize_storage()
 
-        if not self._classkey in self._data.keys():
+        if self._classkey not in self._data.keys():
```
I wonder why `pylint` did not complain here. I can fix it in a separate PR?

adferrand replied:
Yes, that is strange. We heavily modified pylint to avoid it to generate a lot of false positives, maybe it also deactivate some of these kind of nice controls. Definitely it worths to try a fix, in a separate PR I agree.

atombrella reviewed (commented):
There are some areas where I'd like a bit of input, primarily with tests. I think it's maturing; I have some unpushed changes with all of the types of integration tests that were requested in the original PR (at least I think they are).

Searching on cabforum didn't yield a result for those three curves. An updated link/reference would be nice.

atombrella commented on certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:
```diff
@@ -657,4 +732,4 @@ def test_preferred_chain(context):
             assert 'preferred_chain = {}'.format(requested) in f.read(), \
-                   'Expected preferred_chain to be set in renewal config'
\ No newline at end of file
+                   'Expected preferred_chain to be set in renewal config'
```
I think this was the result of my editor.

adferrand replied:
What do you mean here?

alexzorin replied:
When I last modified this file, I failed to include a trailing newline. @atombrella's editor appears to automatically add them.

atombrella replied:
Yes. Exactly. #8297 is an attempt to address some of these settings.

atombrella commented:
@adferrand Can you have a cursory look to see if it's taking shape, and that everything has been addressed? There are probably things in here that should be chopped (or expanded), such as the change in `certbot/certbot/tests/testdata/README`, as I'm not really sure about that part of the test suite.

I couldn't really find valid email addresses so `Co-authored-by` would work probably.

adferrand commented on certbot/CHANGELOG.md:
```diff
@@ -29,6 +29,10 @@ More details about these changes can be found on our GitHub repo.
+* CLI flag `--key-type` has been added to specify 'rsa' or 'ecdsa' (default 'rsa').
+  Only accepts a single value at this time.
+* CLI flag `--elliptic-curve` has been added which takes an NIST/SECG elliptic curve,
+  (see `openssl ecparam -list-curves` for a complete list)
```
In the spirit of the discussion above, we should make clear here that only 3 curves are supported.

atombrella replied:
Yes. I have an unpushed change that will list them :)

adferrand commented on certbot/certbot/crypto_util.py:
```diff
@@ -174,18 +184,39 @@ def import_csr_file(csrfile, data):
-    assert bits >= 1024  # XXX
-    key = crypto.PKey()
-    key.generate_key(crypto.TYPE_RSA, bits)
+    if key_type.lower() == 'rsa':
```
Any reason to lower the value? The message in the error raised in case of wrong value is quite explicit about `rsa` and `ecdsa` being the only acceptable values.

atombrella replied:
Probably this was some of the remnants of the original PR. I'm not really familiar with zope, and how it works. I'll update it.

adferrand commented on certbot/certbot/crypto_util.py:
```diff
@@ -34,16 +35,20 @@
         already exists at the path.
 
-    :param int key_size: RSA key size in bits
+    :param int key_size: key size in bits
```
We should add something like "if key type is rsa".

adferrand commented on certbot/certbot/crypto_util.py:
```diff
@@ -34,16 +35,20 @@
+    :param int key_size: key size in bits
     :param str key_dir: Key save directory.
+    :param str key_type: Key Type [rsa, ecdsa]
+    :param str elliptic_curve: Name of the elliptic curve
```
We should add something like "if key type is ecdsa".

adferrand commented on certbot/certbot/crypto_util.py:
```diff
@@ -174,18 +184,39 @@ def import_csr_file(csrfile, data):
+    """Generate PEM encoded RSA|EC key.
 
-    :param int bits: Number of bits, at least 1024.
+    :param int bits: Number of bits, at least 1024 for RSA.
```
As in `init_save_key`, we should add for `bits` and `ec_curve` that they apply respectively for key type `rsa` and `ecdsa`.

adferrand commented on certbot/certbot/crypto_util.py:
```diff
@@ -174,18 +184,39 @@ def import_csr_file(csrfile, data):
+
+        key = crypto.PKey()
+        key.generate_key(crypto.TYPE_RSA, bits)
+    elif key_type.lower() == 'ecdsa':
```
Same as above, about lowering the value before comparison.

adferrand commented on certbot/certbot/tests/testdata/sample-renewal.conf:
```diff
@@ -44,6 +44,7 @@ apache_handle_sites = True
 authenticator = standalone
 domains = isnot.org,
 rsa_key_size = 2048
+elliptic_curve = secp256k1
```
Should be `secp256kr1`.

adferrand commented on certbot/tests/client_test.py:
```diff
@@ -383,7 +391,9 @@ def test_obtain_certificate_dry_run(self, mock_acme_crypto, mock_crypto):
     @mock.patch("certbot._internal.client.acme_crypto_util")
     def test_obtain_certificate_dry_run_authz_deactivations_failed(self, mock_acme_crypto,
-                                                                   mock_crypto, mock_log):
+                                                                 mock_crypto, mock_log):
```
Alignement error here.

adferrand commented on certbot/tests/crypto_util_test.py:
```diff
@@ -174,11 +175,55 @@ def test_bad_csr(self):
         OpenSSL.crypto.load_privatekey(
-            OpenSSL.crypto.FILETYPE_PEM, make_key(1024))
+            OpenSSL.crypto.FILETYPE_PEM, make_key(elliptic_curve="secp256r1", key_type='ecdsa'))
+        # test the private key
```
This inline comment can be removed.

adferrand commented on tests/letstest/requirements.txt:
```diff
@@ -4,7 +4,7 @@ boto3==1.14.11
 cffi==1.14.0
 colorama==0.4.3
-cryptography==2.8
+cryptography==3.1
```
Do we need specific capabilities from cryptography 3.x ?

atombrella replied:
Probably no. I think it was testing around. I've reverted it.

adferrand commented on letsencrypt-auto-source/tests/auto_test.py:
```diff
@@ -1,6 +1,10 @@
 """Tests for letsencrypt-auto"""
 
-from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
+# PY2,PY3
```
I am surprised that this modification is needed, it does not seem related to the PR.

atombrella commented on acme/acme/crypto_util.py:
```diff
@@ -7,8 +7,7 @@
 import josepy as jose
-from OpenSSL import crypto
-from OpenSSL import SSL  # type: ignore # https://github.com/python/typeshed/issues/2052
+from OpenSSL import crypto, SSL  # type: ignore # https://github.com/python/typeshed/issues/2052
```
I'll check if this change is necessary. I think the linter was complaining about this.

adferrand commented on certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:
```diff
@@ -421,20 +426,79 @@ def test_reuse_key(context):
+    ])
+
+    key1 = join(context.config_dir, "archive", certname, 'privkey1.pem')
+    assert 200 < os.stat(key1).st_size < 250  # ec keys of 256 bits are ~225 bytes
```
I suppose we can use `assert_elliptic_key(key2, SECP256R1)` here.

atombrella replied:
Ah, yes. I added it as an additional assert, just like the other one :) I don't know if the size asserts are needed, but I don't think they do any harm.

adferrand commented on certbot/CHANGELOG.md:
```diff
@@ -29,8 +29,12 @@ More details about these changes can be found on our GitHub repo.
 ### Added
 
-* Added the ability to remove email and phone contact information from an account 
+* Added the ability to remove email and phone contact information from an account
```
Could you move the new changelog entries to the 1.9.0 section?

atombrella replied:
Done :)

atombrella replied:
I had nothing to do with the email ability, it's just my editor deleting trailing whitespace.

atombrella commented:
@adferrand I think all the feedback has been addressed. Please note that I allowed maintainers to push to this branch in case you want to fix something before merging.

adferrand commented:
Indeed @atombrella, the PR is ready to be merged in my opinion. Thanks a lot for your work and you patience regarding the various requirements on that feature. 

I am just waiting for #7694 to be fixed, it should happen in the next days.

osirisinferi commented on certbot/certbot/crypto_util.py:
```diff
@@ -174,18 +185,39 @@ def import_csr_file(csrfile, data):
-    key = crypto.PKey()
-    key.generate_key(crypto.TYPE_RSA, bits)
+    if key_type == 'rsa':
+        if bits < 1024 or bits % 1024 != 0:
```
I know [boulder recently added this restriction](https://github.com/letsencrypt/boulder/pull/4839) b/c of [Debian Weak Keys](https://community.letsencrypt.org/t/issuing-for-common-rsa-key-sizes-only/133839). However, shouldn't certbot be able to generate RSA keys with modulo 8 in stead of 1024, to keep all options open? AFAIK certbot isn't a Let's Encrypt specific ACME client, correct? Other CA's might choose *not* to restrict RSA keys to modulo 1024.

atombrella replied:
@osirisinferi @adferrand Perhaps it's actually better to do this restriction in a separate PR? I agree that 2048 would probably be a good lower limit for the key size. RSA768 in the RSA factoring challenge has been completed, and 1024 is not far from that.

adferrand replied:
Indeed, whether to implement this feature and with which appropriate best practices require some discussions that are outside of the scope of this PR. We can open an issue to work on this.

osirisinferi replied:
@atombrella I didn't mean the 1024 lower limit, I meant the *modulo* 1024 which isn't present in the current certbot code.

adferrand replied:
Indeed I was also talking about the modulo, which would need a proper migration path for users that are not using standard key sizes.

Sadly also we need to revert back to the `assert` control on the key size. Indeed `assert` is operating only during the debug, so technically certbot is not supposed to fail at that point if you ask for a 512bits RSA key (even if it is bad and I am almost sure Boulder an any other decent CA server will refuse it).

But definitely we should discuss that outside of the PR and try to infer good practices through certbot, as it is done usually.

atombrella replied:
OK. I removed the `% 1024` part. I don't really understand why `assert` is a better choice than the error. Personally, I prefer explicit error handling, rather than having `assert` statements tossed in. `import this` has "Explicit is better than implicit."

adferrand replied:
To be clear: I also think an error is better than assert. However, since it was an assert before this PR, the code will break only during development, not on user side. If we change that to a proper exception, the code will break in all situations, which changes certbot behavior. In particular if some people are using short rsa keys, certbot would start to break unexpectedly.

I think we should definitely incite to use decent keys length, by refusing short keys in Certbot. However this needs a proper migration path.

adferrand commented:
Block on #8343

Ruaphoc commented:
I see the PR for #8343 is awaiting review. Is there any chance someone with write access can review and approve both #8343  and this PR for #8254 ? This PR addresses the top priority item that is not currently on the core certbot dev priorities list. It was mentioned in https://community.letsencrypt.org/t/certbot-support-for-ecdsa-certificates/132857/20 as an item that the community could help with, and needed some refinement. I see above the refinement has been completed, and it would greatly help my certbot integration work if this could be merged and released in the very near future.

atombrella commented:
> I see the PR for #8343 is awaiting review. Is there any chance someone with write access can review and approve both #8343 and this PR for #8254 ? This PR addresses the top priority item that is not currently on the core certbot dev priorities list. It was mentioned in https://community.letsencrypt.org/t/certbot-support-for-ecdsa-certificates/132857/20 as an item that the community could help with, and needed some refinement. I see above the refinement has been completed, and it would greatly help my certbot integration work if this could be merged and released in the very near future.

I think it doesn't matter so much in which order these two PRs are merged, as long as they both land in the same release. Perhaps this could be a good compromise? Based on the feedback, this PR seems to have been approved; however, maybe @bmw or @alexzorin would have some final words. I'm sorry to be a bit pushy :)

bmw commented:
This PR and https://github.com/certbot/certbot/pull/8343 are certainly on our radar and I expect both will be merged in the next couple weeks. Thanks again for working on this!

atombrella commented:
I'm trying to add a test similar to the one added in #8343 for the ecdsa key. I guess that'd be one last necessity before the PR will be merged?

adferrand commented:
For that PR, it is. We need to keep the flags associated to the key type and the curve chosen, and have a unit test covering it.

Now there are two other elements, one related to flags interactions, the other related to documentation for key migration, that need to be implemented according to the high level design. We decided to integrate them after this PR, and put this PR on a dedicated branch to have all the features merged together in master by the next release. This release will happen in one month. I will open the relevant issues tonight.

Based on these issues, if you are willing to implement them, it will be much appreciated!

atombrella commented:
> For that PR, it is. We need to keep the flags associated to the key type and the curve chosen, and have a unit test covering it.

Good. The test is currently in my local branch, and when I get it running, I'll push it. I added a new `elif` for `EllipticCurvePrivateKey`.

> Now there are two other elements, one related to flags interactions, the other related to documentation for key migration, that need to be implemented according to the high level design. We decided to integrate them after this PR, and put this PR on a dedicated branch to have all the features merged together in master by the next release. This release will happen in one month. I will open the relevant issues tonight.

Please reference them in this PR, so I don't miss them. Thank you!

> Based on these issues, if you are willing to implement them, it will be much appreciated!

Yes. My goal was to get this PR merged, so yes, I'd at least make an effort at starting, and asking for help if I get stuck.

adferrand commented:
@atombrella I linked the two issues related to the improvements to add on top of this PR.

J0WI commented on certbot/certbot/_internal/cli/__init__.py:
```diff
@@ -314,6 +314,16 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):
+        "security", "--elliptic-curve", type=str, choices=[
+              'secp256r1',
+              'secp384r1',
+              'secp521r1',
```
Note that `openssl ec_params -list-curves` might not be complete due https://github.com/openssl/openssl/issues/7920. Support for `Ed25519` has already been requested in https://github.com/certbot/certbot/issues/2625#issuecomment-325371621.

osirisinferi replied:
I'm not sure if you'd want to have a curve already "active" while it's not allowed by the BR? I can imagine someone adding it to the Boulder code, but not enable it in production until it's actually allowed. That way, it would be possible to enable it relatively quickly.
But certbot doesn't work with configurations the same way. Everything you code here is immediately active and usable by users. It could be confusing for users if they use it in certbot, but get an error from Boulder (or other public ACME endpoints).
Also, if you *do* decide to add Ed25519 to the code, don't forget Ed448. It's in [RFC 8446](https://tools.ietf.org/html/rfc8446) too.

atombrella replied:
@osirisinferi It's in my local branch. I'll push when I get the renewal test working :+1:

Ruaphoc replied:
It wouldn't hurt to add them, but until Boulder supports them, have it throw an error or warning that they are not yet implemented in the ACME back end. When Boulder is updated to support it, remove the error/warning. If it's in certbot and throwing an error/warning that is easy to understand, someone will open an issue or PR to have it enabled once Boulder supports it.

adferrand replied:
I am in favor of using the most conservative set of curves, and add new ones with a proper error handling in a future PR.

atombrella replied:
OK. I've removed them. If possible, I'll have a look at the golang boulder code once this PR has been merged.

atombrella commented:
@adferrand I've addressed the design document, and had initially overseen the "Key Type" for the certificates command. I've added a line for it, but I don't see any code coverage for the command. Adding it is doable, but maybe the PR is growing a bit in size.

The documentation tasks are a bit "tricky" for me to address. I've made an attempt at one of them, and the other one is a bit confusing to me. The one about commands on already existing ecdsa certificates; can you guide a bit, and possibly suggest a section where this would be a good fit to write something about it?

I took out parts about ed448 and ed25519. Maybe it'd be a good opportunity to toy a bit with Golang when this PR has been merged, to add support for those two key types.

Thank you!

TaaviE commented:
Is there anything blocking this PR from being merged and those topics from being discussed elsewhere?

adferrand commented on certbot/docs/using.rst:
```diff
@@ -531,11 +549,11 @@ rate limit.)
-RSA public key. If a certificate is successfully renewed using
-specified options, those options will be saved and used for future
-renewals of that certificate.
+``certbot renew --key-type ecdsa --elliptic-curve secp384r1`` would try
```
I prefer to keep the RSA example for now, since it is the default setup.

adferrand commented on certbot/tests/renewal_test.py:
```diff
@@ -1,4 +1,5 @@
 """Tests for certbot._internal.renewal"""
+import logging
```
I think `logging` is not used here.

adferrand reviewed (changes requested):
Some remaining comments and this is good!

Once done, could you reopen the PR against the `ecdsa` branch in Certbot repository that I just created (it is up to date with `master`)? This way we can implement the remaining tasks, then merge everything in one go to the main branch.

atombrella commented:
Closing, and will re-open shortly in a separate PR.
